tools/lsp/language/formatting.rs

```diff
@@ -0,0 +1,125 @@
+use super::DocumentCache;
+use crate::util::map_range;
+use dissimilar::Chunk;
+use i_slint_compiler::parser::SyntaxToken;
+use lsp_types::{DocumentFormattingParams, TextEdit};
+use rowan::{TextRange, TextSize};
+
+struct StringWriter {
+    text: String,
+}
+
+impl slint_fmt::writer::TokenWriter for StringWriter {
+    fn no_change(&mut self, token: SyntaxToken) -> std::io::Result<()> {
+        self.text += &token.text();
+        Ok(())
+    }
+
+    fn with_new_content(&mut self, _token: SyntaxToken, contents: &str) -> std::io::Result<()> {
+        self.text += contents;
+        Ok(())
+    }
+
+    fn insert_before(&mut self, token: SyntaxToken, contents: &str) -> std::io::Result<()> {
+        self.text += contents;
+        self.text += &token.text();
+        Ok(())
+    }
+}
+
+pub fn format_document(
+    params: DocumentFormattingParams,
+    document_cache: &DocumentCache,
+) -> Option<Vec<TextEdit>> {
+    let file_path = super::uri_to_file(&params.text_document.uri)?;
+    let doc = document_cache.documents.get_document(&file_path)?;
+    let doc = doc.node.as_ref()?;
+
+    let mut writer = StringWriter { text: String::new() };
+    slint_fmt::fmt::format_document(doc.clone(), &mut writer).ok()?;
+
+    let original: String = doc.text().into();
+    let diff = dissimilar::diff(&original, &writer.text);
```

Comment:
(I was hoping that slint-fmt could generate diff directly instead of doing a diff afterwards, but for now this is a perfectly valid way to do it)